src/main/kotlin/com/exactpro/th2/read/file/common/impl/DefaultFileReader.kt

```diff
@@ -31,41 +36,43 @@ import com.exactpro.th2.read.file.common.StreamId
 import com.exactpro.th2.read.file.common.cfg.CommonFileReaderConfiguration
 import com.exactpro.th2.read.file.common.extensions.attributes
 import com.exactpro.th2.read.file.common.state.ReaderState
+import com.exactpro.th2.read.file.common.state.StreamData
 import com.exactpro.th2.read.file.common.state.impl.InMemoryReaderState
+import com.google.protobuf.TextFormat.shortDebugString
 import java.nio.file.Files
 import java.nio.file.Path
 import java.time.Duration
 import java.time.Instant
+import org.apache.commons.lang3.builder.ToStringBuilder
 import kotlin.math.abs
 
-class DefaultFileReader<T : AutoCloseable> private constructor(
+abstract class DefaultFileReader<T : AutoCloseable, MESSAGE_BUILDER, ID_BUILDER> protected constructor(
     configuration: CommonFileReaderConfiguration,
     directoryChecker: DirectoryChecker,
-    contentParser: ContentParser<T>,
+    contentParser: ContentParser<T, MESSAGE_BUILDER>,
     readerState: ReaderState,
-    readerListener: ReaderListener,
-    private val delegateHolder: DelegateHolder<T>,
+    readerListener: ReaderListener<MESSAGE_BUILDER>,
     private val sourceFactory: (StreamId, Path) -> FileSourceWrapper<T>,
-    helper: FileReaderHelper,
-) : AbstractFileReader<T>(
+    helper: FileReaderHelper<MESSAGE_BUILDER, ID_BUILDER>,
+) : AbstractFileReader<T, MESSAGE_BUILDER, ID_BUILDER>(
     configuration,
     directoryChecker,
     contentParser,
     readerState,
     readerListener,
     helper,
 ) {
-    override fun canReadRightNow(holder: FileHolder<T>, staleTimeout: Duration): Boolean = delegateHolder.canRead(holder, staleTimeout)
+    private lateinit var delegateHolder: DelegateHolder<T, MESSAGE_BUILDER>
```

Comment:
Let's try to move this back to the constructor...
And we can modify the `fun createInstance(): DefaultFileReader<T, MESSAGE_BUILDER, MESSAGE_ID>` method to `fun createInstance(delegateHolder: DelegateHolder<T, MESSAGE_BUILDER>): DefaultFileReader<T, MESSAGE_BUILDER, MESSAGE_ID>`

Reply:
Cannot be dome because `DelegateHolder` is a protected class in another class